File: M4 Intermediate DSA/src/L14_Sorting/Assignment3.java
package L14_Sorting;

import java.util.ArrayList;

public class Assignment3 {

  int countFactors(int A) {
    int count = 0;

    for (int num = 1; num * num <= A; num++) {
      if (A % num == 0) {
        count +=num * num == A ? 1 : 2;
      }
    }

    return count;
  }

  public ArrayList<Integer> solve(ArrayList<Integer> A) {

    A.sort((obj1 ,  obj2) -> {
      int aFactor = countFactors(obj1);
      int bFactor = countFactors(obj2);

      return aFactor == bFactor ? obj1 - obj2 : aFactor - bFactor;
    });

    return A;
  }
}

File: M4_Intermediate_DSA/L3_Sorting/assignment3.go
package main

import "sort"

func countFactors(A int) int {
	count := 0

	for num := 1; num*num <= A; num++ {
		if A%num == 0 {
			if num*num == A {
				count++
			} else {
				count += 2
			}
		}
	}
	return count
}

func SolveA3(A []int) []int {

	sort.Slice(A, func(i, j int) bool {
		factor1 := countFactors(A[i])
		factor2 := countFactors(A[j])

		if factor1 < factor2 {
			return true
		} else if factor1 == factor2 && A[i] < A[j] {
			return true
		}

		return false
	})

	return A
}
